README.md

# flutter_screen_scaling
[![pub package](https://img.shields.io/pub/v/flutter_screen_scaling.svg)](https://pub.dartlang.org/packages/flutter_screen_scaling)

** A flutter plugin for adapting screen and font size to different scales.  Originally forked from flutter_screenutil with some cosmetic changes.

[中文文档](https://github.com/SunnyApp/flutter_screen_scaling/blob/master/README_CN.md)  

[github](https://github.com/SunnyApp/flutter_screen_scaling)

[Update log](https://github.com/SunnyApp/flutter_screen_scaling/blob/master/CHANGELOG.md)

## Usage:

### Add dependency：
Please check the latest version before installation.
If there is any problem with the new version, please use the previous version
```
dependencies:
  flutter:
    sdk: flutter
 
  flutter_screen_scaling: ^2.3.0
```

### Add the following imports to your Dart code:
```
import 'package:flutter_screen_scaling/flutter_screen_scaling.dart';
```

### Property
   
|Property|Type|Default Value|Description|
|:---|:---|:---|:---| 
|width|double|1080px|The width of the device in the design draft, in px|
|height|double|1920px|The height of the device in the design draft, in px|
|allowFontScaling|bool|false|Sets whether the font size is scaled according to the system's "font size" assist option|

### Initialize and set the fit size and font size to scale according to the system's "font size" accessibility option
Please set the width and height of the design draft before use, the width and height of the design draft (unit px).
Be sure to set the page in the MaterialApp's home/initialRoute(ie the entry file, just set it once) to ensure that the fit size is set before each use:

```dart

//fill in the screen size of the device in the design

//default value : width : 1080px , height:1920px , allowFontScaling:false
ScreenUtil.init(context);

//If the design is based on the size of the iPhone6 ​​(iPhone6 ​​750*1334)
ScreenUtil.init(context, width: 750, height: 1334);

//If you want to set the font size is scaled according to the system's "font size" assist option
ScreenUtil.init(context, width: 750, height: 1334, allowFontScaling: true);

```

### Use：

### API

#### Pass the px size of the design draft

```dart
    ScreenUtil().setWidth(540)  (sdk>=2.6 : 540.w) //Adapted to screen width
    ScreenUtil().setHeight(200) (sdk>=2.6 : 200.h) //Adapted to screen height
    ScreenUtil().setSp(24)      (sdk>=2.6 : 24.sp)  //Adapter font
    ScreenUtil().setSp(24, allowFontScalingSelf: true)  (sdk>=2.6 : 24.ssp) //Adapter font(fonts will scale to respect Text Size accessibility settings)
    ScreenUtil().setSp(24, allowFontScalingSelf: false)  (sdk>=2.6 : 24.nsp) //Adapter font(fonts will not scale to respect Text Size accessibility settings)

    ScreenUtil.pixelRatio       //Device pixel density
    ScreenUtil.screenWidth   (sdk>=2.6 : 1.wp)    //Device width
    ScreenUtil.screenHeight  (sdk>=2.6 : 1.hp)    //Device height
    ScreenUtil.bottomBarHeight  //Bottom safe zone distance, suitable for buttons with full screen
    ScreenUtil.statusBarHeight  //Status bar height , Notch will be higher Unit px
    ScreenUtil.textScaleFactor  //System font scaling factor

    ScreenUtil().scaleWidth //Ratio of actual width dp to design draft px
    ScreenUtil().scaleHeight //Ratio of actual height dp to design draft px

    0.2.wp  //0.2 times the screen width
    0.5.hp  //50% of screen width
```

#### Adapt screen size：

Pass the px size of the design draft：

Adapted to screen width: `ScreenUtil().setWidth(540)`,

Adapted to screen height: `ScreenUtil().setHeight(200)`,

If your dart sdk>=2.6, you can use extension functions:

example:

instead of :
```
Container(
width: ScreenUtil().setWidth(50),
height:ScreenUtil().setHeight(200),
)
```
you can use it like this:
```
Container(
width: 50.w,
height:200.h
)
```
**Note** 

Height is also adapted according to setWidth to ensure no deformation (when you want a square) 

setHeight method is mainly adapted in height, you want to control the height and actuality of a screen on the UIUsed when the same is displayed.

```dart
//for example:
//rectangle
Container(
           width: ScreenUtil().setWidth(375),
           height: ScreenUtil().setHeight(200),
           ...
            ),
            
////If you want to display a square:
Container(
           width: ScreenUtil().setWidth(300),
           height: ScreenUtil().setWidth(300),
            ),
            
```

#### Adapter font:
``` dart
//Incoming font size，the unit is pixel, fonts will not scale to respect Text Size accessibility settings
//(AllowallowFontScaling when initializing ScreenUtil)
ScreenUtil().setSp(28)    
     
//Incoming font size，the unit is pixel，fonts will scale to respect Text Size accessibility settings
//(If somewhere does not follow the global allowFontScaling setting)
ScreenUtil().setSp(24, allowFontScalingSelf: true)

//for example:

Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Text(
                    'My font size is 24px on the design draft and will not change with the system.',
                    style: TextStyle(
                      color: Colors.black,
                      fontSize: ScreenUtil().setSp(24),
                    )),
                Text(
                    'My font size is 24px on the design draft and will change with the system.',
                    style: TextStyle(
                        color: Colors.black,
                        fontSize: ScreenUtil()
                            .setSp(24, allowFontScalingSelf: true))),
              ],
            )
```

```dart
//import
import 'package:flutter/material.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter_ScreenUtil',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    //Set the fit size (fill in the screen size of the device in the design) If the design is based on the size of the iPhone6 ​​(iPhone6 ​​750*1334)
    ScreenUtil.init(width: 750, height: 1334, allowFontScaling: false);

    return ExampleWidget(title: 'FlutterScreenUtil Demo');
  }
}

class ExampleWidget extends StatefulWidget {
  const ExampleWidget({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _ExampleWidgetState createState() => _ExampleWidgetState();
}

class _ExampleWidgetState extends State<ExampleWidget> {
  @override
  Widget build(BuildContext context) {
    printScreenInformation();
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Row(
              children: <Widget>[
                Container(
                  padding: EdgeInsets.all(ScreenUtil().setWidth(10)),
                  width: ScreenUtil().setWidth(375),
                  height: ScreenUtil().setHeight(200),
                  color: Colors.red,
                  child: Text(
                    'My width:${ScreenUtil().setWidth(375)}dp \n'
                    'My height:${ScreenUtil().setHeight(200)}dp',
                    style: TextStyle(
                        color: Colors.white, fontSize: ScreenUtil().setSp(24)),
                  ),
                ),
                Container(
                  padding: EdgeInsets.all(ScreenUtil().setWidth(10)),
                  width: ScreenUtil().setWidth(375),
                  height: ScreenUtil().setHeight(200),
                  color: Colors.blue,
                  child: Text(
                      'My width:${ScreenUtil().setWidth(375)}dp \n'
                      'My height:${ScreenUtil().setHeight(200)}dp',
                      style: TextStyle(
                          color: Colors.white,
                          fontSize: ScreenUtil().setSp(24))),
                ),
              ],
            ),
            Text('Device width:${ScreenUtil.screenWidth}dp'),
            Text('Device height:${ScreenUtil.screenHeight}dp'),
            Text('Device width:${ScreenUtil.screenWidthPx}px'),
            Text('Device height:${ScreenUtil.screenHeightPx}px'),
            Text('Device pixel density:${ScreenUtil.pixelRatio}'),
            Text('Bottom safe zone distance:${ScreenUtil.bottomBarHeight}dp'),
            Text('Status bar height:${ScreenUtil.statusBarHeight}dp'),
            Text(
              'Ratio of actual width dp to design draft px:${ScreenUtil().scaleWidth}',
              textAlign: TextAlign.center,
            ),
            Text(
              'Ratio of actual height dp to design draft px:${ScreenUtil().scaleHeight}',
              textAlign: TextAlign.center,
            ),
            SizedBox(
              height: ScreenUtil().setHeight(100),
            ),
            Text('System font scaling factor:${ScreenUtil.textScaleFactor}'),
            Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Text(
                    'My font size is 24px on the design draft and will not change with the system.',
                    style: TextStyle(
                      color: Colors.black,
                      fontSize: ScreenUtil().setSp(24),
                    )),
                Text(
                    'My font size is 24px on the design draft and will change with the system.',
                    style: TextStyle(
                        color: Colors.black,
                        fontSize: ScreenUtil()
                            .setSp(24, allowFontScalingSelf: true))),
              ],
            )
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.title),
        onPressed: () {
          ScreenUtil.init(width: 1500, height: 1334, allowFontScaling: false);
          setState(() {});
        },
      ),
    );
  }

  void printScreenInformation() {
    print('Device width dp:${ScreenUtil.screenWidth}'); //Device width
    print('Device height dp:${ScreenUtil.screenHeight}'); //Device height
    print(
        'Device pixel density:${ScreenUtil.pixelRatio}'); //Device pixel density
    print(
        'Bottom safe zone distance dp:${ScreenUtil.bottomBarHeight}'); //Bottom safe zone distance，suitable for buttons with full screen
    print(
        'Status bar height dp:${ScreenUtil.statusBarHeight}dp'); //Status bar height , Notch will be higher Unit dp
    print(
        'Ratio of actual width dp to design draft px:${ScreenUtil().scaleWidth}');
    print(
        'Ratio of actual height dp to design draft px:${ScreenUtil().scaleHeight}');
    print(
        'The ratio of font and width to the size of the design:${ScreenUtil().scaleWidth * ScreenUtil.pixelRatio}');
    print(
        'The ratio of  height width to the size of the design:${ScreenUtil().scaleHeight * ScreenUtil.pixelRatio}');
  }
}

```

### example:

[example demo](/example/lib/main.dart)
 
effect:

![effect](demo_en.png)
![tablet effect](demo_tablet_en.png)


CHANGELOG.md
## [3.3.2+3]
 * Fixing analyzer warning
README

## [3.3.2+2]
 * Rm gemfile
Update version

## [3.3.2+1]
 * Adding build scriptss
null-safety
Generated file

<!--
 * @Author: zhuoyuan93@gmail.com
 * @Date: 2018-10-16 19:43:03
 * @LastEditors: zhuoyuan93@gmail.com
 * @LastEditTime: 2020年6月20日 11:20:02
 * @Description: Update log
 -->

# 2.3.2 (ericm)
- Removing nndb
 
# 2.3.1 (ericm)
- Downgrading sdk version 

# 2.3.0
- We still need context to initialize, sorry.

# 2.2.0
- add 'wp','hp'. Get the height/width of the screen proportionally
- For example: 0.5.wp : Half the width of the screen.

# 2.1.0 
- add 'nsp' , you can use 'fontSize: 24.nsp' instead of 'fontSize: ScreenUtil().setSp(24, allowFontScalingSelf: false)'

# 2.0.0
- Use `MediaQueryData.fromWindow(window)` instead of `MediaQuery.of(context)`, no context parameter required
- API changes, please note

# 1.1.0
- support ExtensionMethod Dart-SDK-2.6.0
- you can use 'width: 50.w' instead of 'width: ScreenUtil().setWidth(50)'
  '50.h' instead of 'ScreenUtil().setHeight(50)'
  '24.sp' instead of 'ScreenUtil().setSp(24)'
  '24.ssp' instead of 'ScreenUtil().setSp(24, allowFontScalingSelf: true)'

# 1.0.2
- fix #89 
- 优化屏幕旋转效果
- 字体适配统一使用宽度

# 1.0.1
- Rebuild code, change API
  Delete "getInstance()", please use "ScreenUtil ()" instead of "ScreenUtil.getInstance()"
  use "ScreenUtil().setSp(24, allowFontScalingSelf: true)" instead of "ScreenUtil.getInstance().setSp(14, true)"
- Modify the initialization method
- Fix #68
- Change example code
  Example CompileSdkVersion change to 28

**If there is significant impact, please return to 0.7.0**

# 0.7.0

- Replace textScaleFactory with textScaleFactor , It's a typo.

# 0.6.1

- Add return types to all methods.

# 0.6.0

- Completing comments , adding English commentsWelcome to add, correct
- 参数同时支持传入 int / double 或者是var size = 100 , var size = 100.0.
- The argument also supports passing in in / double / var size = 100 /var size = 100.0

# 0.5.3

- Change the units of statusBarHeight and bottomBarHeight to dp

# 0.5.2

- Change the parameter type from int to double

- setWidth,setHeight,setSp. for example: you can use setWidth(100) or setWidth(100.0)

# 0.5.1

- Fix the wrong way of using

- It is recommended to use `ScreenUtil.getInstance()` instead of `ScreenUtil()` , for example: `ScreenUtil.getInstance().setHeight(25)` instead of `ScreenUtil().setHeight(25)`

# 0.4.4

- Fix bugs that default fonts change with the system

# 0.4.3

- Modify the font to change with the system zoom mode. The default value is false.

- setSp(int fontSize, [allowFontScaling = false]) => allowFontScaling
? setWidth(fontSize) \* \_textScaleFactor
: setWidth(fontSize);

# 0.4.2

- add two Properties
- ///Current device width dp
- ///当前设备宽度 dp
- ScreenUtil.screenWidthDp

- ///Current device height dp
- ///当前设备高度 dp
- ScreenUtil.screenHeightDp

# 0.4.1

- Fix font adaptation issues

# 0.4.0

- Optimize font adaptation method

# 0.3.1

- Perfect documentation
- Width is enlarged relative to the design draft => The ratio of font and width to the size of the design
- Height is enlarged relative to the design draft => The ratio of height width to the size of the design


# 0.3.0

- Add font size adaptation

# 0.2.2

- Optimize documentation

# 0.0.2

- Fixed bug when releasing

# 0.0.1

- first version






lib/screenutil.dart
/*
 * Created by 李卓原 on 2018/9/29.
 * email: zhuoyuan93@gmail.com
 */

import 'package:flutter/material.dart';

ScreenProperties _instance;
const int screenDefaultWidth = 1080;
const int screenDefaultHeight = 1920;

class ScreenProperties {
  /// UI设计中手机尺寸 , px
  /// Size of the phone in UI Design , px
  final num uiWidthPx;
  final num uiHeightPx;

  /// 控制字体是否要根据系统的“字体大小”辅助选项来进行缩放。默认值为false。
  /// allowFontScaling Specifies whether fonts should scale to respect Text Size accessibility settings. The default is false.
  final bool allowFontScaling;

  final double screenWidth;
  final double screenHeight;
  final double pixelRatio;
  final double statusBarHeight;
  final double bottomBarHeight;
  final double textScaleFactor;

  final double scaleWidth;
  final double scaleHeight;

  ScreenProperties.instance(
      {this.uiWidthPx,
      this.uiHeightPx,
      this.allowFontScaling,
      this.screenWidth,
      this.screenHeight,
      this.pixelRatio,
      this.statusBarHeight,
      this.bottomBarHeight,
      this.textScaleFactor})
      : assert(uiWidthPx != null),
        assert(uiHeightPx != null),
        scaleWidth =
            uiWidthPx == null ? 1 : screenWidth / (uiWidthPx * pixelRatio),
        scaleHeight =
            uiHeightPx == null ? 1 : screenHeight / (uiHeightPx * pixelRatio);

  factory ScreenProperties(
      {double width, double height, bool allowFontScaling}) {
    if (width != null || height != null || allowFontScaling != null) {
      return _instance = ScreenProperties.instance(
        uiWidthPx: width ?? WidgetsBinding.instance.window.physicalSize.width,
        uiHeightPx:
            height ?? WidgetsBinding.instance.window.physicalSize.height,
        allowFontScaling: allowFontScaling,
        pixelRatio: WidgetsBinding.instance.window.devicePixelRatio,
        screenWidth: WidgetsBinding.instance.window.physicalSize.width,
        screenHeight: WidgetsBinding.instance.window.physicalSize.height,
        statusBarHeight: WidgetsBinding.instance.window.padding.top,
        bottomBarHeight: WidgetsBinding.instance.window.padding.bottom,
        textScaleFactor: WidgetsBinding.instance.window.textScaleFactor,
      );
    } else {
      if (_instance == null) {
        throw "Not initialized";
      }
      return _instance;
    }
  }

  double get scaleText => scaleWidth;

  /// 根据UI设计的设备宽度适配
  /// 高度也可以根据这个来做适配可以保证不变形,比如你先要一个正方形的时候.
  /// Adapted to the device width of the UI Design.
  /// Height can also be adapted according to this to ensure no deformation ,
  /// if you want a square
  num convertWidth(num width) => width * scaleWidth;

  /// 根据UI设计的设备高度适配
  /// 当发现UI设计中的一屏显示的与当前样式效果不符合时,
  /// 或者形状有差异时,建议使用此方法实现高度适配.
  /// 高度适配主要针对想根据UI设计的一屏展示一样的效果
  /// Highly adaptable to the device according to UI Design
  /// It is recommended to use this method to achieve a high degree of adaptation
  /// when it is found that one screen in the UI design
  /// does not match the current style effect, or if there is a difference in shape.
  num convertHeight(num height) => height * scaleHeight;

  ///字体大小适配方法
  ///@param [fontSize] UI设计上字体的大小,单位px.
  ///Font size adaptation method
  ///@param [fontSize] The size of the font on the UI design, in px.
  ///@param [allowFontScaling]
  num convertFontSize(num fontSize, {bool allowFontScaling}) {
    allowFontScaling ??= this.allowFontScaling ?? false;
    return allowFontScaling
        ? (fontSize * scaleText)
        : ((fontSize * scaleText) / textScaleFactor);
  }
}

class ScreenUtil {
  ScreenUtil._();
  factory ScreenUtil({double width, double height, bool allowFontScaling}) {
    ScreenProperties(
        width: width, height: height, allowFontScaling: allowFontScaling);
    return ScreenUtil._();
  }

  /// 每个逻辑像素的字体像素数，字体的缩放比例
  /// The number of font pixels for each logical pixel.
  static double get textScaleFactor => instance.textScaleFactor;

  /// 设备的像素密度
  /// The size of the media in logical pixels (e.g, the size of the screen).
  static double get pixelRatio => instance.pixelRatio;

  static ScreenProperties get instance {
    if (_instance == null) {
      throw "You must initialize ScreenUtil first";
    }
    return _instance;
  }

  /// 当前设备宽度 dp
  /// The horizontal extent of this size.
  static double get screenWidth => instance.screenWidth;

  ///当前设备高度 dp
  ///The vertical extent of this size. dp
  static double get screenHeight => instance.screenHeight;

  /// 当前设备宽度 px
  /// The vertical extent of this size. px
  static double get screenWidthPx => instance.screenWidth * instance.pixelRatio;

  /// 当前设备高度 px
  /// The vertical extent of this size. px
  static double get screenHeightPx =>
      instance.screenHeight * instance.pixelRatio;

  /// 状态栏高度 dp 刘海屏会更高
  /// The offset from the top
  static double get statusBarHeight => instance.statusBarHeight;

  /// 底部安全区距离 dp
  /// The offset from the bottom.
  static double get bottomBarHeight => instance.bottomBarHeight;
}


example/lib/main.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter_screen_scaling/flutter_screen_scaling.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Screen Scaling',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    //Set the fit size (fill in the screen size of the device in the design) If the design is based on the size of the iPhone6 ​​(iPhone6 ​​750*1334)
    ScreenScaleProperties(
        width: 750, height: 1334, allowFontScaling: false, allowSubpixel: true);

    return ExampleWidget(title: 'Screen Scaling Demo');
  }
}

class ExampleWidget extends StatefulWidget {
  const ExampleWidget({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _ExampleWidgetState createState() => _ExampleWidgetState();
}

class _ExampleWidgetState extends State<ExampleWidget> {
  @override
  Widget build(BuildContext context) {
    //printScreenInformation();
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Row(
              children: <Widget>[
                Container(
                  padding: EdgeInsets.all(ScreenScale.convertWidth(10)),
                  width: 375.w,
                  height: 200.h,
                  color: Colors.red,
                  child: Text(
                    'My width:${375.w}dp \n'
                    'My height:${200.h}dp',
                    style: TextStyle(
                        color: Colors.white,
                        fontSize: ScreenScale.convertFontSize(24)),
                  ),
                ),
                Container(
                  padding: EdgeInsets.all(ScreenScale.convertWidth(10)),
                  width: ScreenScale.convertWidth(375),
                  height: ScreenScale.convertHeight(200),
                  color: Colors.blue,
                  child: Text(
                      'My width:${0.5.wp}dp \n'
                      'My height:${ScreenScale.convertHeight(200)}dp',
                      style: TextStyle(
                          color: Colors.white,
                          fontSize: ScreenScale.convertFontSize(24))),
                ),
              ],
            ),
            Text('Device width:${ScreenScale.screenWidthPx}px'),
            Text('Device height:${ScreenScale.screenHeightPx}px'),
            Text('Device width:${ScreenScale.screenWidth}dp'),
            Text('Device height:${ScreenScale.screenHeight}dp'),
            Text('Device pixel density:${ScreenScale.pixelRatio}'),
            Text('Bottom safe zone distance:${ScreenScale.bottomBarHeight}dp'),
            Text('Status bar height:${ScreenScale.statusBarHeight}dp'),
            Text(
              'Ratio of actual width dp to design draft px:${ScreenScale.scaleWidth}',
              textAlign: TextAlign.center,
            ),
            Text(
              'Ratio of actual height dp to design draft px:${ScreenScale.scaleHeight}',
              textAlign: TextAlign.center,
            ),
            SizedBox(
              height: ScreenScale.convertHeight(100),
            ),
            Text('System font scaling factor:${ScreenScale.textScaleFactor}'),
            Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Text(
                    'My font size is 24px on the design draft and will not change with the system.',
                    style: TextStyle(
                      color: Colors.black,
                      fontSize: 24.px,
                    )),
                Text(
                    'My font size is 24px on the design draft and will change with the system.',
                    style: TextStyle(
                        color: Colors.black,
                        fontSize: ScreenScale.convertFontSize(24,
                            allowFontScaling: true))),
              ],
            )
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.title),
        onPressed: () {
          ScreenScale.init(
            width: 1500,
            height: 1334,
            allowFontScaling: false,
            allowSubpixel: true,
          );
          setState(() {});
        },
      ),
    );
  }

  void printScreenInformation() {
    print('Device width dp:${ScreenScale.screenWidth}'); //Device width
    print('Device height dp:${ScreenScale.screenHeight}'); //Device height
    print(
        'Device pixel density:${ScreenScale.pixelRatio}'); //Device pixel density
    print(
        'Bottom safe zone distance dp:${ScreenScale.bottomBarHeight}'); //Bottom safe zone distance，suitable for buttons with full screen
    print(
        'Status bar height px:${ScreenScale.statusBarHeight}dp'); //Status bar height , Notch will be higher Unit px
    print(
        'Ratio of actual width dp to design draft px:${ScreenScale.scaleWidth}');
    print(
        'Ratio of actual height dp to design draft px:${ScreenScale.scaleHeight}');
    print(
        'The ratio of font and width to the size of the design:${ScreenScale.scaleWidth * ScreenScale.pixelRatio}');
    print(
        'The ratio of  height width to the size of the design:${ScreenScale.scaleHeight * ScreenScale.pixelRatio}');
    print('System font scaling:${ScreenScale.textScaleFactor}');
    print('0.5 times the screen width:${0.5.wp}');
    print('0.5 times the screen height:${0.5.hp}');
  }
}


lib/size_extension.dart
import 'dart:math' as math;

import 'screen_scale_properties.dart';

void sizeExtensions() {}

class Px {}

extension SizeExtension on num {
  ///[ScreenScale.setWidth]
  double get w => ScreenScale.convertWidth(this).roundTo(3);

  ///[ScreenScale.setHeight]
  double get h => ScreenScale.convertHeight(this).roundTo(3);

  ///[ScreenScale.setSp]
  double get px => ScreenScale.convertFontSize(this).roundTo(3);

  ///[ScreenScale.setSp]
  double get ptScaled =>
      ScreenScale.convertFontSize(this, allowFontScaling: true).roundTo(3);

  ///[ScreenScale.setSp]
  double get ptNoScaling =>
      ScreenScale.convertFontSize(this, allowFontScaling: false).roundTo(3);

  ///屏幕宽度的倍数
  ///Multiple of screen width
  double get wp => (ScreenScale.screenWidth * this).roundTo(3);

  ///屏幕高度的倍数
  ///Multiple of screen height
  double get hp => (ScreenScale.screenHeight * this).roundTo(3);

  double roundTo([int? precision]) {
    var mod = math.pow(10.0, precision!);
    return ((this * mod).round().toDouble() / mod);
  }
}


example/lib/main_zh.dart
import 'package:flutter/material.dart';
import 'package:flutter_screen_scaling/flutter_screen_scaling.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter_ScreenUtil',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: "Flutter Screen Util"),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    //设置适配尺寸 (填入设计稿中设备的屏幕尺寸) 此处假如设计稿是按iPhone6的尺寸设计的(iPhone6 750*1334)

    ScreenScaleProperties(
        width: 750, height: 1334, allowFontScaling: false, allowSubpixel: true);

    return ExampleWidget(title: 'FlutterScreenUtil示例');
  }
}

class ExampleWidget extends StatefulWidget {
  const ExampleWidget({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _ExampleWidgetState createState() => _ExampleWidgetState();
}

class _ExampleWidgetState extends State<ExampleWidget> {
  @override
  Widget build(BuildContext context) {
    printScreenInformation();
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Row(
              children: <Widget>[
                Container(
                  padding: EdgeInsets.all(ScreenScale.convertWidth(10)),
                  width: ScreenScale.convertWidth(375),
                  height: ScreenScale.convertHeight(200),
                  color: Colors.red,
                  child: Text(
                    '我的宽度:${0.5.wp}dp \n'
                    '我的高度:${ScreenScale.convertHeight(200)}dp',
                    style: TextStyle(
                        color: Colors.white,
                        fontSize: ScreenScale.convertFontSize(24)),
                  ),
                ),
                Container(
                  padding: EdgeInsets.all(ScreenScale.convertWidth(10)),
                  width: 375.w,
                  height: 200.h,
                  color: Colors.blue,
                  child: Text(
                      '我的宽度:${375.w}dp \n'
                      '我的高度:${200.h}dp',
                      style: TextStyle(
                          color: Colors.white,
                          fontSize: ScreenScale.convertFontSize(24))),
                ),
              ],
            ),
            Text('设备宽度:${ScreenScale.screenWidthPx}px'),
            Text('设备高度:${ScreenScale.screenHeightPx}px'),
            Text('设备宽度:${ScreenScale.screenWidth}dp'),
            Text('设备高度:${ScreenScale.screenHeight}dp'),
            Text('设备的像素密度:${ScreenScale.pixelRatio}'),
            Text('底部安全区距离:${ScreenScale.bottomBarHeight}dp'),
            Text('状态栏高度:${ScreenScale.statusBarHeight}dp'),
            Text(
              '实际宽度的dp与设计稿px的比例:${ScreenScale.scaleWidth}',
              textAlign: TextAlign.center,
            ),
            Text(
              '实际高度的dp与设计稿px的比例:${ScreenScale.scaleHeight}',
              textAlign: TextAlign.center,
            ),
            SizedBox(
              height: 100.h,
            ),
            Text('系统的字体缩放比例:${ScreenScale.textScaleFactor}'),
            Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Text('我的文字大小在设计稿上是24px，不会随着系统的文字缩放比例变化',
                    style: TextStyle(
                      color: Colors.black,
                      fontSize: 24.px,
                    )),
                Text('我的文字大小在设计稿上是24px，会随着系统的文字缩放比例变化',
                    style: TextStyle(
                        color: Colors.black,
                        fontSize: ScreenScale.convertFontSize(24,
                            allowFontScaling: true))),
              ],
            )
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.title),
        onPressed: () {
          ScreenScaleProperties(
              width: 1500,
              height: 1334,
              allowFontScaling: false,
              allowSubpixel: true);
          setState(() {});
        },
      ),
    );
  }

  void printScreenInformation() {
    print('设备宽度:${ScreenScale.screenWidth}'); //Device width
    print('设备高度:${ScreenScale.screenHeight}'); //Device height
    print('设备的像素密度:${ScreenScale.pixelRatio}'); //Device pixel density
    print(
        '底部安全区距离:${ScreenScale.bottomBarHeight}dp'); //Bottom safe zone distance，suitable for buttons with full screen
    print(
        '状态栏高度:${ScreenScale.statusBarHeight}dp'); //Status bar height , Notch will be higher Unit px

    print('实际宽度的dp与设计稿px的比例:${ScreenScale.scaleWidth}');
    print('实际高度的dp与设计稿px的比例:${ScreenScale.scaleHeight}');

    print(
        '宽度和字体相对于设计稿放大的比例:${ScreenScale.scaleWidth * ScreenScale.pixelRatio}');
    print('高度相对于设计稿放大的比例:${ScreenScale.scaleHeight * ScreenScale.pixelRatio}');
    print('系统的字体缩放比例:${ScreenScale.textScaleFactor}');

    print('屏幕宽度的0.5:${0.5.wp}');
    print('屏幕高度的0.5:${0.5.hp}');
  }
}
